gcovr/csv_generator.py

```diff
@@ -0,0 +1,47 @@
+# -*- coding:utf-8 -*-
+
+# This file is part of gcovr <http://gcovr.com/>.
+#
+# Copyright 2013-2018 the gcovr authors
+# Copyright 2013 Sandia Corporation
+# This software is distributed under the BSD license.
+
+import sys
+import csv
+
+from .utils import calculate_coverage, sort_coverage, presentable_filename
+
+
+def print_csv_report(covdata, output_file, options):
+    """produce gcovr csv report"""
+    if output_file:
+        with open(output_file, 'w') as fh:
+            _real_print_csv_report(covdata, fh, options)
+    else:
+        _real_print_csv_report(covdata, sys.stdout, options)
+
+
+def _real_print_csv_report(covdata, OUTPUT, options):
+    total_lines = 0
+    total_covered = 0
+
+    # Data
+    keys = sort_coverage(
+        covdata, show_branch=options.show_branch,
+        by_num_uncovered=options.sort_uncovered,
+        by_percent_uncovered=options.sort_percent)
+
+    def _summarize_file_coverage(coverage):
+        filename = presentable_filename(
+            coverage.filename, root_filter=options.root_filter)
+
+        if options.show_branch:
+            total, cover, percent = coverage.branch_coverage()
+        else:
+            total, cover, percent = coverage.line_coverage()
+        return (filename, cover, total, percent)
+
+    writer = csv.writer(OUTPUT)
+    writer.writerow('filename', 'covered', 'total', 'percent')
+    for key in keys:
+        writer.writerow(_summarize_file_coverage(covdata[key]))
```

Comment:
instead of switching between line/branch coverage summaries, couldn't we show all? I'd add the following columns:

* filename
* lines covered
* lines total
* lines percent
* branches covered
* branches total
* branches percent

We should also be explicit about how the percent fields are represented. If it's called percent, does it have a `%` sign? Or do we show fractions like 1/10 = 10% = `0.1`?

Reply:
all good points. push-forced fixes to export all of those, and changed the percentages to [0,1.0] range